src/NCronJob/Scheduler/TaskScheduler/ConcurrencyTaskScheduler.cs

```diff
@@ -0,0 +1,69 @@
+using System.Collections.Concurrent;
+
+namespace NCronJob;
+
+/// <summary>
+/// Controlled parallelism and balanced concurrency: Use when you need to manage the maximum degree of parallelism explicitly
+/// and ensure tasks are processed with a controlled level of concurrency.
+/// </summary>
+internal class ConcurrencyTaskScheduler : TaskScheduler
+{
+    [ThreadStatic]
+    private static bool currentThreadIsProcessingItems;
+
+    private readonly ConcurrentQueue<Task> tasks = new();
+    private readonly int maxDegreeOfParallelism;
+    private int delegatesQueuedOrRunning;
+
+    public ConcurrencyTaskScheduler(int maxDegreeOfParallelism)
+    {
+        ArgumentOutOfRangeException.ThrowIfLessThan(maxDegreeOfParallelism, 1);
+
+        this.maxDegreeOfParallelism = maxDegreeOfParallelism;
+    }
+
+    protected sealed override void QueueTask(Task task)
+    {
+        tasks.Enqueue(task);
+        if (Interlocked.CompareExchange(ref delegatesQueuedOrRunning, maxDegreeOfParallelism, maxDegreeOfParallelism) < maxDegreeOfParallelism)
+        {
+            Interlocked.Increment(ref delegatesQueuedOrRunning);
+            NotifyThreadPoolOfPendingWork();
+        }
+    }
+
+#pragma warning disable CA2008
+    private void NotifyThreadPoolOfPendingWork() =>
+        ThreadPool.UnsafeQueueUserWorkItem(_ =>
```

Comment:
Is there a reason not to use `QueueUserWorkItem`?
Given that this code can be exploitable: https://learn.microsoft.com/en-us/dotnet/api/system.threading.threadpool.unsafequeueuserworkitem?view=net-8.0#system-threading-threadpool-unsafequeueuserworkitem(system-threading-waitcallback-system-object)

Reply:
This was really more of a test to observe what happens during profiling.
The three custom TaskScheduler are still experimental. Currently it's only using the Deterministic one.

Reply:
That said, we may want to squash **some** of the commits :D - given this beauty:

> [Hopefully we don't merge that](https://github.com/NCronJob-Dev/NCronJob/pull/85/commits/a53c9b0879bc5287098f148238abee6ab33ebfff)